src/rgw/rgw_reshard.cc

```diff
@@ -513,10 +531,14 @@ static int commit_reshard(rgw::sal::RadosStore* store,
 
   int ret = fault.check("commit_target_layout");
   if (ret == 0) { // no fault injected, write the bucket instance metadata
-    ret =
-      store->getRados()->put_bucket_instance_info(bucket_info, false,
-						  real_time(),
-						  &bucket_attrs, dpp);
+    int tries = 0;
+    do {
+      ret =
+	store->getRados()->put_bucket_instance_info(bucket_info, false,
+						    real_time(),
+						    &bucket_attrs, dpp);
+      ++tries;
+    } while (ret == -ECANCELED && tries < max_retries);
```

Comment:
if we detect ECANCELED from this write, we'll need to re-read the `bucket_info` from rados. a lot of the earlier logic in `commit_reshard()` is based on values from `bucket_info`, so should be reevaluated inside of this retry loop also

we'll also need to check for races with `cancel_reshard()` and subsequent reshards. the easiest thing to do here would be to bail if `bucket_info.layout != prev`, but that would catch benign races with bilog trimming of `layout.logs`

so i think we should be more specific, and only check for changes to `layout.resharding`, `layout.current_index`, or `layout.target_index`

the strategy for `revert_target_layout()` should be very similar